sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -125,16 +125,13 @@ object DataSourceV2Strategy extends Strategy {
       val filterCondition = postScanFilters.reduceLeftOption(And)
       val withFilter = filterCondition.map(FilterExec(_, scan)).getOrElse(scan)
 
-      val withProjection = if (withFilter.output != project) {
-        ProjectExec(project, withFilter)
-      } else {
-        withFilter
-      }
-
-      withProjection :: Nil
+      // always add the projection, which will produce unsafe rows required by some operators
+      ProjectExec(project, withFilter) :: Nil
 
     case r: StreamingDataSourceV2Relation =>
-      DataSourceV2ScanExec(r.output, r.source, r.options, r.pushedFilters, r.reader) :: Nil
+      // ensure there is a projection, which will produce unsafe rows required by some operators
+      ProjectExec(r.output,
```

Comment:
continuous streaming scan always return unsafe rows, will we introduce regression here?  cc @jose-torres

Reply:
For now I think it's safer to still require `DataSourceV2ScanExec` to return unsafe rows and move the unsafe conversion to here in a followup PR.

Reply:
Continuous processing will still be experimental in the 2.4 release, so I'm not tremendously concerned about this. We should eventually change the scan to produce rows in whatever way is most efficient in the final API.

Reply:
It is perfectly fine for sources to produce UnsafeRow because it is an InternalRow.

I think it is important for us to get to InternalRow in this release. UnsafeRow is too hard to produce and the easiest thing to do is to produce InternalRow and then call into Spark's UnsafeProjection to produce UnsafeRow. That's painful, uses internal APIs, and is slower.